lib/board.js

```diff
@@ -8,28 +8,67 @@ var Board = function(options) {
 };
 
 Board.prototype.getClickedTile = function(clickedX, clickedY) {
-  let tileArray = this.tiles.filter(function(tile) {
-    return clickWithinTile(tile, clickedX, clickedY);
-  });
-  return tileArray[0];
+  return this.tiles.find(tile => clickWithinTile(tile, clickedX, clickedY));
 };
 
 Board.prototype.getTurnTiles = function() {
-  return this.tiles.filter(function(tile) {
-    return tile.constructor.name === 'PathTile' && tile.isTurnPoint;
-  });
+  return this.tiles.filter(tile => tile.isPathTile() && tile.isTurnPoint);
+};
+
+Board.prototype.getHighlightedOrSelectedBuildTiles = function() {
+  return this.tiles.filter(tile => tile.isHighlightedBuildTile() ||
+                                   tile.isSelectedBuildTile());
 };
 
 Board.prototype.removeAllTowers = function() {
-  this.getTilesWithTowers().forEach(function(tile) {
-    tile.removeTower();
-  });
+  this.getTilesWithTowers().forEach(tile => tile.removeTower());
 };
 
 Board.prototype.getTilesWithTowers = function() {
-  return this.tiles.filter(function(tile) {
-    return tile.type === 'buildTile' && !tile.vacant;
-  });
+  return this.tiles.filter(tile => tile.isBuildTile() && tile.hasTower());
+};
+
+Board.prototype.setCurrentTile = function(x, y) {
+  this.lastTile = this.currentTile;
+  this.currentTile = this.getClickedTile(x, y);
+};
+
+Board.prototype.currentTileIsAVacantBuildTile = function() {
+  return this.currentTile &&
+         this.currentTile.isBuildTile() &&
+         this.currentTile.vacant;
+};
+
+Board.prototype.lastTileIsHighlightedOrSelected = function() {
+  return this.lastTile &&
+         (this.lastTile.isHighlightedBuildTile() ||
+          this.lastTile.isSelectedBuildTile());
+};
+
+Board.prototype.currentTileIsABuildTile = function() {
+  return this.currentTile && this.currentTile.isBuildTile();
+};
+
+Board.prototype.currentTileIsASelectedBuildTile = function() {
+  return this.currentTile && this.currentTile.isSelectedBuildTile();
+};
+
+Board.prototype.currentTileIsAHighlightedBuildTile = function() {
+  return this.currentTile && this.currentTile.isHighlightedBuildTile();
+};
+
+Board.prototype.updateTileTypes = function () {
```

Comment:
@stevekinney I moved a whole lot of logic out of engine.js but it's resulted in a mess in the Board class. However, I'm not sure how to get around this mess because I have to keep checking if there is a currently clicked tile/previously clicked tile (`currentTile` and `lastTile`) and there's a lot of different logic based on what type of tiles those are. Any suggestions on how to get out of this mess??